### PR TITLE
remove images from metadata on source update

### LINKED ISSUE #15635: Remove unused attachments from cell metadata
Images pasted into cells are added to cell metadata as attachments and are rendered as such in markdown code. If the user deletes the image, it deletes it from the code but it doesn't remove the attachment from the metadata. 

Provide a way to remove the unused attachments from polluting the metadata.

### PR DESCRIPTION
**Pull request by MaddyDev**
This PR partly addresses #15635 
When user deletes image from the rich text mode, it deletes it from metadata.
Note - Removing attachments when user deletes them from markdown code cell is still pending to be implemented.

### CODE REVIEW

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
This breaks images when making edits in split/MD mode : 

1. Paste an image in WYSIWYG mode
2. Switch to split/MD mode
3. Make any change (type a character)

The attachments are cleared and now since the source doesn't have the image stream directly in the source anymore it doesn't re-add them and so the attachment links are broken.

**Comment by MaddyDev** [comment]
> This breaks images when making edits in split/MD mode :
> 
> 1. Paste an image in WYSIWYG mode
> 2. Switch to split/MD mode
> 3. Make any change (type a character)
> 
> The attachments are cleared and now since the source doesn't have the image stream directly in the source anymore it doesn't re-add them and so the attachment links are broken.

Thanks @Charles-Gagnon! I was only testing the rich text scenarios, totally missed this one!

File: src/sql/workbench/services/notebook/browser/models/cell.ts [hunk 1]
```diff
@@ -149,6 +149,10 @@ export class CellModel extends Disposable implements ICellModel {
 		return this._attachments;
 	}
 
+	public set attachments(imageData: nb.ICellAttachments) {
+		this._attachments = imageData ?? {};
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/cell.ts:153]: imageData isn't really accurate - attachments can be many different things. Would suggest just using `attachments` as the param name.

And the type should be `nb.ICellAttachments | undefined`

File: src/sql/workbench/services/notebook/browser/models/modelInterfaces.ts [hunk 2]
```diff
@@ -533,7 +533,7 @@ export interface ICellModel {
 	sendChangeToNotebook(change: NotebookChangeType): void;
 	cellSourceChanged: boolean;
 	readonly savedConnectionName: string | undefined;
-	readonly attachments: nb.ICellAttachments;
+	attachments: nb.ICellAttachments;
```
Comment (Charles-Gagnon) [inline src/sql/workbench/services/notebook/browser/models/modelInterfaces.ts:536]: This should be `nb.ICellAttachments | undefined` if we're following what the actual implementation is saying. (and then also please update the _attachments property in the cell class)

**Review by Charles-Gagnon — APPROVED** [review]
This looks fine for fixing the rich text view, but what about deleting images in split/MD mode? I think those situations could be fixed by going through the attachments after making a cell source change and then removing any that aren't referenced in the Notebook.

But maybe we don't want to do that - that might get annoying if you have an attachment you reference and then want to move it, so cut out the reference text to paste elsewhere and then the attachment itself is deleted because of that....so I'd suggest following up with design on the correct approach here. 

I would strongly suggest adding some tests though - we really want to make sure that attachments are working correctly and given how complicated the code is with the multiple different views relying on manual tests alone is going to be error prone I think. (But you should also add manual tests to the test plan as well if you haven't already)

**Comment by Charles-Gagnon** [comment]
> Thanks @Charles-Gagnon! I was only testing the rich text scenarios, totally missed this one!

This sounds like an(other) excellent reason to add tests for these scenarios 😄

**Comment by Charles-Gagnon** [comment]
@chlafreniere I'm going to merge this in so that the next release can have this fix, then if you get anyone else who can work on this stuff they can follow up with the questions I asked above regarding split/MD view

**Pull request merged**